rmw_fastrtps_cpp/src/TypeSupport.cpp

```diff
@@ -202,8 +202,10 @@ bool TypeSupport::serializeROSmessage(eprosima::fastcdr::Cdr &ser, const rosidl_
             {
                 case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_BOOL:
                     {
-                        bool sb = false, &b = *(bool*)field;
-                        if(b) sb = true; 
+                        bool sb = false;
+                        // don't cast to bool here because if the bool is
+                        // uninitialized the random value can't be deserialized
+                        if(*(uint8_t*)field) sb = true;
```

Comment:
Basically it comes down to uninitialized values result in undefined behavior. One example see http://markshroyer.com/2012/06/c-both-true-and-false/ In this specific case it looks like that the compiler generated assembler code makes assumptions about the value being 0 or 1 (since it is supported to be bool) which leads to sb actually taking the value of field (since the compiler reuses a register which is supposed to contain the intended value.